### PR TITLE
RFC: rtio: SPI and a Fast Sensor

### PR DESCRIPTION
**Pull request by teburd**
Many changes in this RFC PR are very unlikely to make it upstream as-is or perhaps ever but as a practical demonstration for how RTIO could be used with a sensor, on a spi bus it does just that.

The sample app itself (samples/boards/tdk_robokit1) shows how delays may result in buffer underruns/overruns (depending on the perspective!), and how they may be resolved by having more buffers to feed the demand of the sensor. It would be interesting to also consider a throughput/latency comparison... lots of small buffers vs a few large ones.

It also lets you build with and without CONFIG_SPI_RTIO to show how thread stacks might be saved by using RTIO with SPI. Once #55178 is merged I'll rebase again and we can do probably even more compare/contrasts to get a sense of things.

This builds on top of #55382

### CODE REVIEW

File: include/zephyr/drivers/spi.h [hunk 1]
```diff
@@ -878,6 +918,117 @@ __deprecated static inline int spi_write_async(const struct device *dev,
...
+{
+	const struct spi_dt_spec *dt_spec = iodev_sqe->iodev->data
+	const struct device *dev = dt_spec->bus;
+	const struct spi_driver_api *api = (const struct spi_driver_api *)bus->api;
+
+	return api->iodev_poll(dt_spec->bus, iodev_sqe);
+}
+
+/**
+ * @brief Convert a transcieve from spi_buf_sets to a sequence of RTIO ops
```
Comment (tbursztyka) [inline include/zephyr/drivers/spi.h:940]: s/transceive/transceive  

Typo seem to repeat in many place below.

Comment (teburd) [reply include/zephyr/drivers/spi.h:940]: Thank you, yes I'm sure there's many other things in here that will need correction :-)

**Comment by teburd** [comment]
Patch set is mostly cleaned up now and worth perusing for criticisms and such.

File: include/zephyr/rtio/rtio.h [hunk 2]
```diff
@@ -133,20 +134,34 @@ struct rtio_sqe {
...
+		struct {
+			uint8_t tiny_buf_len; /**< Length of tiny buffer */
+			uint8_t tiny_buf[7]; /**< Tiny buffer */
+		};
+
+		/** OP_CALLBACK */
+		struct {
+			void (*callback)(struct rtio *r, struct rtio_sqe *sqe, void *arg0);
+			void *arg0;
+		};
```
Comment (MaureenHelm) [inline include/zephyr/rtio/rtio.h]: Is this change related to `rtio: Add tiny write op`?

File: include/zephyr/drivers/spi.h [hunk 3]
```diff
@@ -878,6 +892,47 @@ __deprecated static inline int spi_write_async(const struct device *dev,
...
+ * These do not need to be shared globally but doing so
+ * will reduce a small amount of memory usage is the preferred
+ * method.
+ *
+ * @param node DT_NODE
+ */
+#define SPI_IODEV_DEFINE(name, node_id, operation_, delay_)			\
+	const struct spi_dt_spec _spi_dt_spec_##name =				\
+		SPI_DT_SPEC_GET(node_id, operation_, delay_);			\
+	RTIO_IODEV_DEFINE(name, &spi_iodev_api, (void *)&_spi_dt_spec_##name)
```
Comment (MaureenHelm) [inline include/zephyr/drivers/spi.h:932]: `SPI_IODEV_DT_DEFINE` ?

File: include/zephyr/drivers/sensor.h [hunk 4]
```diff
@@ -593,6 +634,70 @@ static inline int z_impl_sensor_channel_get(const struct device *dev,
...
+ * @brief Submit a request for the fifo stream (iodev) for this sensor.
+ */
+static inline int sensor_fifo_iodev(const struct device *dev, struct rtio_iodev **iodev)
+{
+#ifdef CONFIG_RTIO
+
+	const struct sensor_driver_api *api =
+		(const struct sensor_driver_api *)dev->api;
+
+	return api->fifo_iodev(dev, iodev);
```
Comment (MaureenHelm) [inline include/zephyr/drivers/sensor.h:647]: Might want to check if `fifo_iodev` is NULL

File: drivers/sensor/icm42688/Kconfig [hunk 5]
```diff
@@ -4,6 +4,8 @@
 #
 # SPDX-License-Identifier: Apache-2.0
 
+DT_COMPAT_ICM42688 := invensense,icm42688
```
Comment (MaureenHelm) [inline drivers/sensor/icm42688/Kconfig:7]: Most (all?) of these have been removed. Is this a rebasing mistake?

**Review by MaureenHelm — COMMENTED** [review]
some nitpicky comments, but overall this looks very cool!

File: drivers/sensor/icm42688/Kconfig (hunk 5 above)
Comment (teburd) [reply drivers/sensor/icm42688/Kconfig:7]: Almost certainly
